compiler/rustc_hir_typeck/src/expr.rs

```diff
@@ -1672,7 +1672,17 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
 
         let mut seen_fields = FxHashMap::default();
 
-        let mut error_happened = false;
+        let mut error_happened = if variant.fields.len() > remaining_fields.len() {
```

Comment:
No need to touch `error_happened` either. Or better yet, you could make `error_happened` into `Option<ErrorGuaranteed>` and then taint the infcx if it's `Some(..)`.